packages/contracts/contracts/voting/token/TokenVoting.sol

```diff
@@ -70,10 +73,10 @@ contract TokenVoting is MajorityVotingBase {
             revert ProposalCreationForbidden(_msgSender());
         }
 
-        proposalId = proposalCount++;
+        id = _createProposal(_msgSender(), _metadata, _actions);
```

Comment:
This pattern seems to me a vice versa of what we're used to. Currently, it's possible that inherriting contract forgets to call `_createProposal` inside it which will be hard to detect at all times by the audit. The pattern should be that `createProposal` is in the abtsrcta contrcat + in the interface which calls `_createProposal` inside. This way, no way to forget to call as it calls behind the hood without user forgetting it or not. I Get the idea why we doing what we doing currently, but can't we find a better way ? ThOugh s? @mathewmeconry @Rekard0

Reply:
There is no way to ensure that a third-party developer emits the events and/or updates the `proposalId` as we want.

The idea in this PR was to provide code that can be used by every governance plugin implementation and that takes away these responsibilities from the user as much as possible. 

On the developer portal and with boilerplates, we can educate and guide people to write theire governance plugin as we intend.

Reply:
IMO, it is an overkill, i think the IProposal is enough, so devs are free to mix `PluginUUPSUpgradeable` + `IProposal` .....etc

but if we must,
I can imagine, one might think; ok it is only for internal use, but as u said it come with these risk of forgetting...etc
so an idea could be, to create a modifiers that happens first and increase the counter and emit event....etc, so that the counter could be used by the next function...

Reply:
Yes, it seems an overkill and it still doesn't give you advantage and safety., I don't know what the plan is, but if you prefer to first merge it and then we refactor it better, i'm good if you guys are.

Reply:
The idea was to prevent code duplication. We would otherwise need to copy the variable declarations and content of the functions
```solidity
using Counters for Counters.Counter;

/// @notice The incremental ID for proposals and executions.
Counters.Counter private proposalCounter;

/// @notice Checks if this or the parent contract supports an interface by its ID.
/// @param interfaceId The ID of the interface.
/// @return bool Returns `true` if the interface is supported.
function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
    return interfaceId == type(IProposal).interfaceId || super.supportsInterface(interfaceId);
}

/// @inheritdoc IProposal
function proposalCount() external view returns (uint256) {
    return proposalCounter.current();
}

/// @inheritdoc GovernanceBase
function _createProposal(
    address _creator,
    bytes calldata _metadata,
    IDAO.Action[] calldata _actions
) internal virtual override returns (uint256 proposalId) {
    proposalId = proposalCounter.current();
    proposalCounter.increment();

    emit ProposalCreated({
        proposalId: proposalId,
        creator: _creator,
        metadata: _metadata,
        actions: _actions
    });
}

/// @inheritdoc GovernanceBase
function _executeProposal(uint256 _proposalId, IDAO.Action[] memory _actions)
    internal
    virtual
    override
{
    bytes[] memory execResults = dao.execute(_proposalId, _actions);
    emit ProposalExecuted({proposalId: _proposalId, execResults: execResults});
}
```
into **every** governance plugin. This is an overkill to me + the work of testing every implementation separately scales with the number of plugins we maintain.
Moreover, keep in mind that abstract base classes don't cost and we obtain a way to discriminate governance plugins from normal ones through the `supportsInterface` method. 

I think the fact that you percieve this as an overkill comes from the fact that 4 contracts had to be created.
Ultimately, this boils down to our `enum PluginType {UUPS, Cloneable, Constructable}` design.
However, this can be reduced to one file if @novaknole would abstract the `dao` variable away from `DaoAuthorizable` (e.g. in https://github.com/aragon/core/pull/155) so that `GovernanceBase` can inherit from it and use `getDAO()` in `executeProposal` instead of the DAO variable.

Then we would need only one additional class.

Reply:
yah i get your concern for "code duplication" but then if plugins move in to their own separate package, and if so, we need to maintain versioning for this shared common interface, and i am not sure if it worth it.
your call guys.

Reply:
This shared common interface will be on the core repository (like the `Plugin`, `PluginClonable`, and `PluginUUPSUpgradeable` contracts).

Reply:
I am leaning more towards the idea of a common interface... and then developers can implement however they prefer. And IMO, the interface should be enforcing the public, recognizable method names like `createProposal(...)`, not `_createProposal`, which only does part of the job.

For recycling logic, we could use a library?

Reply:
As suggested by @Rekard0 @novaknole and @mathewmeconry, I removed the `GovernancePlugin` related contracts and the `IProposal` interafce in 87dc96d3cb401c8d5b939105bac1f801cc53ce4b and duplicated the code for now.
In PR https://github.com/aragon/core/pull/170, I reverted the changes so that my work can use as the basis to refactor the code.

Reply:
> I am leaning more towards the idea of a common interface... and then developers can implement however they prefer. And IMO, the interface should be enforcing the public, recognizable method names like `createProposal(...)`, not `_createProposal`, which only does part of the job.

Because the arguments of `createProposal` are always different for each governance plugin (some require, start and end dates, or special options) is not possible without using `bytes _data` and doing decoding. This would mean that people that want to create proposals through a smart contract composer don't know what the input should be. It also puts more burden on the developer to do this decoding, which he might not want to do because of bad UX.

> For recycling logic, we could use a library?
We could try this and include the proposalId counter in the library.

Reply:
> This would mean that people that want to create proposals through a smart contract composer don't know what the input should be

That's a very good point.

Still I would try to find a cleaner way to expose the public interface with clean names, not internal underscored ones

Reply:
I agree. something needs to be thought of about this, but as we said, it will be done in another PR.  We can resolve this for no, but This discussion is useful here